src/librustc_typeck/rscope.rs

```diff
@@ -50,6 +51,71 @@ pub trait RegionScope {
     /// computing `object_lifetime_default` (in particular, in legacy
     /// modes, it may not be relevant).
     fn base_object_lifetime_default(&self, span: Span) -> ty::Region;
+
+    /// If this scope allows anonymized types, return the generics in
+    /// scope, that anonymized types will close over. This property is
+    /// controlled by the region scope because it's fine-grained enough
+    /// to allow restriction of anonymized types to the syntactical extent
+    /// of a function's return type.
```

Comment:
Are you talking about allowing `impl Trait`, or the "generics in scope"? The latter would be easy to provide (`<'a, T>` for `fn foo<'a, T>(...) -> impl Trait`), but controlling whether `impl Trait` is legal in a certain position is harder to explain.
For it to make more sense, `rscope` would have to become one of several properties that get propagated recursively.

At the same time, the `ItemContext` in `typeck::collect` may be a better choice now for holding the "generics in scope", as it's the only `AstConv` implementation which allows `impl Trait`, but it's still not that uniform (only the return type of a function or inherent impl method supports it, not just any type in the signature of those items).